### PR TITLE
Update OS, actions, node

### PR DESCRIPTION
**Pull request by Spiker985**
- ~Remove node 14~
  - Add maintenance and LTS node versions
- Update `checkout` to 3.3.0
- Update `setup-node` to 3.6.0
- ~Change windows to `latest`~
  - Cannot change windows to latest until we bump the bundled npm

### CODE REVIEW

**Comment by Spiker985** [comment]
Needs #53 to hopefully fix the windows test

**Review by DeeDeeG — COMMENTED** [review]
This would need some changes before being ready to merge.

I tried to explain what is going on with each of these things and why/with enough context to get what all is happening. On reflection, there's a fair amount of hidden jank around this repo, I guess, TBH.

File: .github/workflows/CI.yml
Comment (confused-Techie) [reply .github/workflows/CI.yml:19]: Second this mention here of testing on everything we can. At least this part of testing can be cheap on implementation side

Comment (Spiker985) [reply .github/workflows/CI.yml]: Restored and expanded in commit a2cd2b9. Targets include 14, 16, 18, and 19 so that we are covered for all maintenance releases, LTS, and current.

File: package.json
Comment (Spiker985) [reply package.json:62]: Still in the process of trying to fix the spec so that it does a proper reference without the need for hoisting the `node-gyp` package into the package.lock. Which should also have the added benefit (when completed) of being entirely node version agnostic

File: .github/workflows/CI.yml
Comment (Spiker985) [reply .github/workflows/CI.yml]: Reverted in commit a2cd2b9. I (ironically enough) discovered this myself in https://github.com/Spiker985/x-terminal-reloaded/commit/84a59bb60cccfdb390c5dc524cf31dfc0a6bdcb4

**Comment by Spiker985** [comment]
image
Well there's your problem...

Looks like I'll need to find another way to add a space into that directory lol

File: package.json
Comment (DeeDeeG) [reply package.json:62]: I think we may have a much harder time doing that test with npm 8, which I hope to move to before May, since npm 6 is going EOL in May.

So it may be small returns for fixing this and potentially having to drop the test entirely for npm 8. It's a long story but I can try to explain it if folks want, but basically the npm config `node-gyp`, which gives a path to the copy of node-gyp somewhere on your machine that you want to use, the config is not respected in npm 7, and I don't think it's respected in npm 8 or 9 either?? Would need to test again, my memory is foggy.

But we would basically have to copy-paste (or symlink?) all of ppm (or at least all of npm?) into a new dir with space in the path and run the test that way... Or just drop the test.

An Atom maintainer was leaning toward/comfortable with "just drop it" last I talked to them about it.

Sorry to go on and on, kind of brain dumping about this, feel free to ask questions or do what speaks to you, I want to be more helpful than annoying here. Not trying to impose. Just I have this stuff vaguely in mind for my to-dos already around this repo, but last thing I want to do is put others off contributing, you know?

File: .github/workflows/CI.yml
Comment (DeeDeeG) [inline .github/workflows/CI.yml:7]: It can be valuable to run CI on a WIP branch, so I'd personally like the `push` trigger to stay active for _all branches_, not just `master`.

Comment (Spiker985) [reply .github/workflows/CI.yml:7]: So, as long as a PR is created it'll still run the CI on pushes. It just won't run _both_ whenever a push happens

File: package.json
Comment (Spiker985) [reply package.json:62]: I've gone ahead and made the executive decision to simply comment out that test. If it won't work going forward, then it's just an impedance and I'm comfortable with no longer including it.

**Comment by Spiker985** [comment]
...why in the world is Windows 16 using `node-gyp -v v9.1.0`...don't tell me that having node-gyp hoisted into the package.json is what made Windows install using the correct version :man_facepalming:

**Comment by Spiker985** [comment]
I give up. Putting back the hoisted `node-gyp` entry into the package.json fixed the CI for Windows 16 AND 18 - so I guess it stays there for now...

File: package-lock.json
Comment (Spiker985) [reply package-lock.json]: I reset the package.json, package-lock.json and yarn.lock from `master` so this is no longer a concern - I appreciate the assistance

**Review by confused-Techie — APPROVED** [review]
So I'll preface by saying I'm not nearly as familiar as @DeeDeeG is with this codebase and it's idiosyncrasies. So if they say something different default to their answer.

But from this, this PR looks great to me!

Commenting out a problematic test, bumping the GitHub Actions we use to be properly supported, and adding many more NodeJS versions to run tests on are all super fantastic.

The only thing I'd want to see here to improve this PR, which if done my approval still counts, would be to add a small comment at the start of the commented out tests saying why it's commented out, or that just contains a link to this PR, so that developers in the future can see why it was commented out and don't have to do the legwork of finding out or using git blame to do so.

But fantastic work @Spiker985, lots of awesome effort on this PR, which is great to see!

**Comment by DeeDeeG** [comment]
I worry about micro-managing this PR, my apologies.

Responding to request for review.

Regarding the CI failure on Node 19: Would personally rather not test any non-LTS Node versions.

<details>

I kind of don't want to worry about non-LTS Node. We have plenty of time to integrate those changes after they've been stabilized and packaged for wide consumption, when they appear in Node 20 LTS in the future. There's generally three LTS releases at any given time, that gives us plenty of lead time to get ready for the newest LTS, in my opinion, before being forced to switch. Riding the "Current" release stream is a headache.

</details>

Also, commenting out a test feels kind of wrong, I don't think it is genuinely relevant to the changes in `.github/workflows/CI.yml`. I do expect this test will not work with npm 8 or newer, so its days are most likely numbered, so I suppose removing it here is okay, just seems not neatly scoped to "one PR one change".

And yet we have some much wider-reaching PRs in other repos. So. I dunno if it's so bad...

Again, I don't want to micro-manage, I feel like this started as a draft and I nudged it into being merge-ready by specifying all the details exactly, doesn't feel like good reviewer behavior, but er it was requested now. Feel free to ping on Discord if you need to talk about it/check in expectations in a less formal way.

**Pull request merged**
